File: Java/Building/src/main/java/Building.java

public class Building {
    private String _name;
    private int _floors;
    private int _roomspfloor;
    private final int _windowsproom = 4;
    private boolean _doorslocked;

    public Building(String name, int floors, int roomspfloor){
        this._name = name;
        this._floors = floors;
        this._roomspfloor = roomspfloor;
        this._doorslocked = true;
    }

    public void enterbuilding() throws BuildingErrorState{
        if(!this._doorslocked){
            System.out.println("Entering " + this._name);
        }
        else{
            throw Locked_Building_State;
        }
    }

    public int totalrooms(){return  this._floors * this._roomspfloor; }

    public int totalwindows(){ return totalrooms() * this._windowsproom;}

    public void lockDoors(){ this._doorslocked = true;}
    public void unlockDoors(){ this._doorslocked = false;}
    public boolean getDoorState() {return this._doorslocked;}

    public String get_name(){ return  this._name; }

    static public BuildingErrorState Locked_Building_State = new BuildingErrorState("You are trying to enter a locked building.");
}

File: CPP/src/building/Building.cpp
#include <iostream>
#include "Building.h"

namespace building{
    Building::Building(std::string name, int floors, int roomspfloor) : _floors(floors), _roomspfloor(roomspfloor) {
        _buildingName = name;
    }
    int Building::totalrooms() {
        return _floors * _roomspfloor;
    }
    int Building::totalWindows() {
        return (totalrooms() * _windowsproom);
    }
    void Building::lockDoors(){
        _doorslocked = true;
    }
    void Building::unlockDoors(){
        _doorslocked = false;
    }
    bool Building::getDoorState(){
        return _doorslocked;
    }
    void Building::enterbuilding(){
        if(getDoorState() == false){
            std::cout << "Entering " << get_name() << std::endl;
        }
        else {throw Locked_Building_State;}
    }
    std::string Building::get_name(){
        return _buildingName;
    }

    Building::~Building() {
            std::cout << "destructing building@" 
                        << ((void*) this) << std::endl; 
        }

    const BuildingErrorState Building::Locked_Building_State("The building is locked.");
}

